### PR TITLE
Bring Angular Devtools Into angular/angular

### PR DESCRIPTION
**Pull request by AleksanderBodurri**
This PR brings in the [Angular DevTools](https://github.com/rangle/angular-devtools) codebase into the Angular Repo. It preserves the commit history of the original repository with a process described [here](https://docs.google.com/document/d/1I4F8aKYe285prD3aXnu3yEoW-3FlG5DeuMFXp2fnLME).

The Angular DevTools codebase has been nested under a `devtools` directory. I will be working on integrating DevTools' Unit and E2E tests with the framework repository's CI.

Remaining tasks before merge
- [x] Verify that all authors have signed the CLA (Done)

Tasks After merge
- Reenable commit-message format validator in CI. Disabled because commits from the Angular DevTools repo did not follow the 20 character minimum limit for commits. Had to temporarily disable until these commits are merged.
- Reenable the branch rebasing workflow in CI. Had to temporarily disable because the commit graph from Angular DevTools  and the commit graph for angular/angular would be disjoint except for the unrelated-histories merge commit, meaning rebasing onto master would fail in CI.
- Build devtools with local angular packages

### CODE REVIEW

**Comment by mgechev** [comment]
@googlebot I consent

**Comment by sumitarora** [comment]
@googlebot I consent

**Comment by markwhitfeld** [comment]
@googlebot I consent.

**Comment by AndrewKushnir** [comment]
@AleksanderBodurri thanks for working on this!

We've just discussed this PR during the framework team sync and we are trying to assess whether this code should live in the `angular/angular` repo or it should be located in it's own repo in Angular organization (for ex. `angular/devtools`). We'll have more discussion early next week and let you know. I just wanted to share it sooner in case there is some major work that you plan within the next couple days to integrate CI, etc.

**Comment by kamikopi** [comment]
@googlebot i consent

File: .ng-dev/format.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ export const format: FormatConfig = {
   'clang-format': {
     'matchers': [
       '**/*.{js,ts}',
+      '!devtools/**',
```
Comment (josephperrott) [inline .ng-dev/format.ts:15]: We can do add it to the list of tasks after we merged if you prefer, but we should just run the formatter over all of the files in the devtools area before any work begins on the code once its in angular/angular as now would beb the least disruptive time for us to do so.

File: .pullapprove.yml [hunk 2]
```diff
@@ -1171,6 +1171,25 @@ groups:
...
+      - >
+        contains_any_globs(files, [
+          'devtools/**',
+          ])
+    reviewers:
+      users:
+        - mgechev
+        - twerske
+        - josephperrott
+        - devversion
```
Comment (josephperrott) [inline .pullapprove.yml:1191]: @mgechev Can you confirm the list of reviewers you were thinking for ownership here?

File: .circleci/config.yml [hunk 3]
```diff
@@ -727,6 +733,30 @@ jobs:
           name: Test all windows CI targets
           command: bazel test --test_tag_filters="-browser:chromium-local" //packages/compiler-cli/...
           no_output_timeout: 15m
+  test_angular_devtools:
+    executor:
+      name: default-executor
```
Comment (josephperrott) [inline .circleci/config.yml:738]: Rather than using the default executor and then manually installing the karma and cypress deps.  We should just create another executor definition here that uses the `cypress/browsers` image (ideally on v16 of node)

**Review by josephperrott — COMMENTED** [review]
Overall I think we are in good shape, I left a few comments about a couple things.  But we should be good to land this in the next few days I think.  As you noted in the description there is a list of cleanup items to do after it lands, but I think we are in good shape for landing it.

**Review by josephperrott — CHANGES_REQUESTED** [review]
One change we will need before we can land it though is that all of the the typescript files will need to contain the correct license header in them.

**Comment by josephperrott** [comment]
To ensure the license header values are present, should add devtools to the tslint scope.

You can add it as included in `/tsconfig-tslint.json`

File: .pullapprove.yml (hunk 2 above)
Comment (mgechev) [reply .pullapprove.yml:1191]: I'd suggest also adding @AleksanderBodurri

**Comment by ladraum** [comment]
@googlebot i consent

File: devtools/.circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+version: 2.1
```
Comment (devversion) [inline devtools/.circleci/config.yml:1]: Is there an item in your cleanup list to remove this file?

Comment (devversion) [reply devtools/.circleci/config.yml:1]: I see a lot of such files that should be removed, like the issue template etc. I assume though this is planned as a follow-up?

File: devtools/WORKSPACE [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    package_json = ["//:package.json"],
+)
+
+# Fetch sass rules for compiling sass files
+http_archive(
+    name = "io_bazel_rules_sass",
+    patch_args = ["-p1"],
+    patches = [
+        # Updates @bazel/work dep to 4.0.0 inside rules_sass so it is compatible
+        "//:io_bazel_rules_sass.patch",
```
Comment (devversion) [inline devtools/WORKSPACE:50]: This workaround is no longer needed and can be removed. Angular components no longer does this.

File: devtools/packages.bzl [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# List of MDC packages.
+MDC_PACKAGES = [
```
Comment (devversion) [inline devtools/packages.bzl:2]: These constants can be removed from this sub-workspace. They were specific to Angular Material

File: devtools/packages.bzl [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    ("@angular/common", struct(entry_points = ["http/testing", "http", "testing"])),
+    ("@angular/compiler", struct(entry_points = ["testing"])),
+    ("@angular/core", struct(entry_points = ["testing"])),
+    ("@angular/forms", struct(entry_points = [])),
+    ("@angular/platform-browser", struct(entry_points = ["testing", "animations"])),
+    ("@angular/platform-server", struct(entry_points = [], platform = "node")),
+    ("@angular/platform-browser-dynamic", struct(entry_points = ["testing"])),
+    ("@angular/router", struct(entry_points = [])),
+    ("@angular/localize", struct(entry_points = ["init"])),
+]
```
Comment (devversion) [inline devtools/packages.bzl:48]: Also note that this list should be the minimum for the devtools workspace. The more modules & entry-points listed here, the more builds are needed for the linked ESM bundles.

File: devtools/projects/demo-no-zone/src/app/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,18 @@
+load("//tools:angular_ts_project.bzl", "ng_ts_project")
+
+package(default_visibility = ["//visibility:public"])
+
+ng_ts_project(
```
Comment (devversion) [inline devtools/projects/demo-no-zone/src/app/BUILD.bazel:5]: Is there any reason `ts_project` is chosen over `ts_library`. `ts_library` would have various benefits:

* Consistency with the whole Angular organization
* Strict dependency checking
* Better incrementality as `ts_project` IIRC rebuilds dependant targets even if the types have not changed.

I think this can be a follow-up, but I think it's definitely something worth discussing, or did you already with @josephperrott?

File: devtools/projects/ng-devtools-backend/src/lib/directive-forest/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+karma_web_test_suite(
+    name = "test",
+    deps = [
+        ":test-lib",
+    ],
+)
+
+ts_test_project(
+    name = "test-lib",
```
Comment (devversion) [inline devtools/projects/ng-devtools-backend/src/lib/directive-forest/BUILD.bazel:30]: Just a heads up nit: Usually the convention is to use underscore as delimiter within targets. The only exception is the one matching the current package name (so that e.g. `bazel test devtools/a/b/c/test` would work).

File: devtools/tools/defaults.bzl [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+load("//tools/angular:index.bzl", "LINKER_PROCESSED_FW_PACKAGES")
+
+esbuild = _esbuild
+esbuild_config = _esbuild_config
+
+def karma_web_test_suite(name, **kwargs):
+    web_test_args = {}
+    test_deps = kwargs.get("deps", [])
+
+    kwargs["tags"] = ["partial-compilation-integration"] + kwargs.get("tags", [])
```
Comment (devversion) [inline devtools/tools/defaults.bzl:16]: This can be removed as you do not have any partial compilation integration tests set up. And I don't think that makes sense to do this in this workspace.

File: devtools/tools/defaults.bzl [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        name = "%s_bundle" % name,
+        deps = test_deps,
+        platform = "browser",
+    )
+
+    # Set up default browsers if no explicit `browsers` have been specified.
+    if not hasattr(kwargs, "browsers"):
+        kwargs["tags"] = ["native"] + kwargs.get("tags", [])
+        kwargs["browsers"] = [
+            # Note: when changing the browser names here, also update the "yarn test"
```
Comment (devversion) [inline devtools/tools/defaults.bzl:29]: Nit: this comment only applied in the components repository and could be removed here.

File: devtools/tools/angular/index.bzl [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+        # List of dependencies which should never be bundled into these linker-processed bundles.
+        external = ["rxjs", "@angular", "domino", "xhr2"],
+    )
+
+    _linker_mapping(
+        name = "%s_linked" % target_name_base,
+        srcs = [":%s_linked_bundle" % target_name_base],
+        package = "@npm//@angular/%s" % pkg.name,
+        module_name = module_name,
+        subpath = target_name_base,
```
Comment (devversion) [inline devtools/tools/angular/index.bzl:64]: We should probably move this to the shared dev-infra code in a follow-up. I'll make a note.

The more I think about all of this though, I realize that technically all of this would not be needed if we'd consume the targets from the `angular` workspace directly. These are built in full compilation mode by default, so the linker would not be needed (just as a quick note)

File: devtools/src/environments/BUILD.bazel [hunk 13]
```diff
@@ -0,0 +1,14 @@
+load("//tools:typescript.bzl", "ts_project")
+
+package(default_visibility = ["//visibility:public"])
+
+ts_project(
+    name = "environments",
+    srcs = [
+        "environment.ts",
+    ],
+    tsconfig = "//:tsconfig.json",
```
Comment (devversion) [inline devtools/src/environments/BUILD.bazel:10]: For a follow-up: Isn't the `tsconfig` attribute here and everywhere else unnecessary? you specified a default in the Starlark macro.

**Comment by DaanDeSmedt** [comment]
@googlebot i consent

File: devtools/CODE_OF_CONDUCT.md [hunk 14]
```diff
@@ -0,0 +1,12 @@
+# Contributor Code of Conduct
+## Version 0.3b-angular
+
+As contributors and maintainers of the Angular DevTools project, we pledge to respect everyone who contributes by posting issues, updating documentation, submitting pull requests, providing feedback in comments, and any other activities.
+
+Communication through any of Angular's channels (GitHub, Gitter, IRC, mailing lists, Google+, Twitter, etc.) must be constructive and never resort to personal attacks, trolling, public or private harassment, insults, or other unprofessional conduct.
```
Comment (JimMorrison723) [inline devtools/CODE_OF_CONDUCT.md:6]: i am pretty sure it's okay to remove Google+ from the list...^^

Comment (mgechev) [reply devtools/CODE_OF_CONDUCT.md:6]: We can remove this file completely.

File: .circleci/config.yml [hunk 15]
```diff
@@ -275,6 +277,7 @@ jobs:
           paths:
             - ./ng
             - ./bazel_repository_cache
+            - .cache/Cypress
```
Comment (devversion) [inline .circleci/config.yml:280]: I know this is deeply integrated already so far, but couple of important notes here:

1. This will increase the cache even more, with binaries. CircleCI will throttle the download if it exceeds 500MB IIRC and we were close to exceeding that before/might already do?

2. I think we need to make a discussion whether we'd want to use Cypress in this repo / have all team members install it, especially since it's just rarely used for the majority of people in the repo, right?

3. Cypress does not work with M1 natively and requires Rosseta 2 (which is something I'd like to avoid; since we want to support m1 natively in all repos): https://github.com/cypress-io/cypress/issues/4478.

I think ideally we'd run the Cypress tests within Bazel, and only download Cypress when actually needed, or we use tooling that is used elsewhere in the repo, if that is possible? (not sure if cypress is needed for the devtools)

cc. @josephperrott @mgechev

File: devtools-bazel.rc [hunk 16]
```diff
@@ -0,0 +1,2 @@
+build --workspace_status_command="yarn -s ng-dev release build-env-stamp --mode=snapshot"
+build --stamp
```
Comment (devversion) [inline devtools-bazel.rc:2]: Is this used somewhere?

File: devtools/.bazelignore [hunk 17]
```diff
@@ -0,0 +1,16 @@
+# NB: sematics here are not the same as .gitignore
+# see https://github.com/bazelbuild/bazel/issues/8106
+.git
+bazel-out
```
Comment (devversion) [inline devtools/.bazelignore:4]: IIRC we can remove this file since this is no longer a nested bazel workspace right?

File: devtools/.bazelversion [hunk 18]
```diff
@@ -0,0 +1 @@
+4.0.0
```
Comment (devversion) [inline devtools/.bazelversion:1]: Same for the two files above (can be removed?)

File: devtools/.npmrc [hunk 19]
```diff
@@ -0,0 +1 @@
+scripts-prepend-node-path=true
```
Comment (devversion) [inline devtools/.npmrc:1]: Nit: I think this can be removed as well?

File: devtools/projects/demo-no-zone/.bazelignore [hunk 20]
```diff
@@ -0,0 +1,3 @@
+node_modules
+dist
+bazel-out
```
Comment (devversion) [inline devtools/projects/demo-no-zone/.bazelignore:3]: I think this file can be removed?

File: devtools/projects/demo-no-zone/src/BUILD.bazel [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+dev_server(
+    name = "devserver",
+    srcs = [":dev_app_static_files"],
+    additional_root_paths = ["angular/devtools/projects/demo-no-zone/src/devapp"],
+    tags = ["manual"],
+    deps = [
+        ":devapp",
+    ],
+)
```
Comment (devversion) [inline devtools/projects/demo-no-zone/src/BUILD.bazel:62]: + 💯 for using the custom devserver. I wanted to move away from the concatjs one regardless!

File: devtools/projects/ng-devtools/src/lib/devtools-tabs/profiler/timeline/recording-visualizer/BUILD.bazel [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    name = "recording_visualizer_styles",
+    srcs = [
+        "bar-chart.component.scss",
+        "bargraph-visualizer.component.scss",
+        "execution-details.component.scss",
+        "flamegraph-visualizer.component.scss",
+        "timeline-visualizer.component.scss",
+        "tree-map-visualizer.component.scss",
+    ],
+)
```
Comment (devversion) [inline devtools/projects/ng-devtools/src/lib/devtools-tabs/profiler/timeline/recording-visualizer/BUILD.bazel:16]: For a follow-up to track: We should remove this target in favor of individual `sass_binary` targets. The multi sass binary rule is not very efficient and not used elsewhere in Angular. More details here: https://github.com/bazelbuild/rules_sass#multi_sass_binary.

We can have a custom macro that will build up multiple `sass_binary` directly, although in general explicit targets are preferred for explicit deps etc.

File: devtools/projects/ng-devtools/src/lib/vendor/angular-split/lib/BUILD.bazel [hunk 23]
```diff
@@ -0,0 +1,28 @@
+load("//devtools/tools:ng_module.bzl", "ng_module")
+load("@io_bazel_rules_sass//:defs.bzl", "sass_binary")
+
+package(default_visibility = ["//visibility:public"])
+
+sass_binary(
+    name = "split_component_styles",
+    src = "//devtools/projects/ng-devtools/src/lib/vendor/angular-split/lib/component:split.component.scss",
```
Comment (devversion) [inline devtools/projects/ng-devtools/src/lib/vendor/angular-split/lib/BUILD.bazel]: If `lib/component` is its own Bazel package, why is the sass binary declared outside? (symptomatic for the inconveniently large label here). We could clean this up unless you had a reason for that.

File: devtools/tools/esbuild/esbuild-spec.config.mjs [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+ * The plugin is not loaded at top-level as not all spec bundle targets rely
+ * on the linker and this would slow-down bundling.
+ */
+ async function fetchAndCreateLinkerEsbuildPlugin() {
+  // Note: This needs to be a NPM module path as this ESBuild config is generated and can
+  // end up in arbitrary Bazel packages or differently-named consumer workspaces.
+  const {createLinkerEsbuildPlugin} = await import(
+    '@angular/dev-infra-private/shared-scripts/angular-linker/esbuild-plugin.mjs'
+  );
+  return await createLinkerEsbuildPlugin(/.*/, /* ensureNoPartialDeclaration */ true, { unknownDeclarationVersionHandling: 'ignore' });
```
Comment (devversion) [inline devtools/tools/esbuild/esbuild-spec.config.mjs]: Can you please add a comment on why we ignore version errors? (capturing the messages Joost basically put together I think)

File: devtools/tools/esbuild/BUILD.bazel [hunk 25]
```diff
@@ -0,0 +1,28 @@
+load("//tools:defaults.bzl", "esbuild_config")
+load(":index.bzl", "create_angular_bundle_targets")
+
+package(default_visibility = ["//visibility:public"])
+
+esbuild_config(
+    name = "esbuild_config",
```
Comment (devversion) [inline devtools/tools/esbuild/BUILD.bazel]: Can this name be more specific? e.g. `esbuild_config_esm` or something? It's unclear what this config targets/what is is used for.

File: devtools/tools/esbuild/esbuild-iife.config.mjs [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export default {
+  resolveExtensions: ['.mjs', '.js'],
+  format: 'iife',
+  mainFields: ['es2020', 'es2015', 'module', 'main'],
+  conditions: ['es2020', 'es2015', 'module'],
```
Comment (devversion) [inline devtools/tools/esbuild/esbuild-iife.config.mjs]: You may also want to have the base options in a base config `.mjs` file, instead of repeating the resolve fields/extensions. three times?

File: package.json [hunk 27]
```diff
@@ -34,7 +35,13 @@
...
     "ts-circular-deps:check": "yarn -s ng-dev ts-circular-deps check --config ./packages/circular-deps-test.conf.js",
     "ts-circular-deps:approve": "yarn -s ng-dev ts-circular-deps approve --config ./packages/circular-deps-test.conf.js",
-    "check-tooling-setup": "yarn tsc --project .ng-dev/tsconfig.json"
+    "check-tooling-setup": "yarn tsc --project .ng-dev/tsconfig.json",
+    "devtools:devserver": "ibazel run //devtools/src:devserver",
+    "devtools:build:chrome": "bazelisk build --config snapshot //devtools/projects/shell-chrome/src:prodapp",
+    "devtools:test": "bazelisk test --config snapshot //devtools/...",
+    "devtools:e2e": "start-server-and-test devtools:devserver http-get://localhost:4200 cy:run",
+    "cy:run": "$(yarn bin)/cypress run --config-file=devtools/cypress.json",
+    "cy:open": "$(yarn bin)/cypress open --config-file=devtools/cypress.json"
```
Comment (devversion) [inline package.json]: Can these two scripts be more specific to the devtools (in terms of naming?). Also I think you can just use `cypress` directly as Yarn will expose `.bin` in the `PATH`

File: tslint.json [hunk 28]
```diff
@@ -139,7 +141,12 @@
       "./modules/benchmarks_external/**/*",
       // Ignore zone.js directory
       // TODO(JiaLiPassion): add zone.js back later
-      "./packages/zone.js/**/*"
+      "./packages/zone.js/**/*",
+
+      "./devtools/dist/**/*",
+      "./devtools/bazel-bin/**/*",
+      "./devtools/bazel-out/**/*",
```
Comment (devversion) [inline tslint.json]: will these folders ever exist? I think bazel-bin/out will not since we have a single Bazel workspace in this repo

File: tools/BUILD.bazel [hunk 29]
```diff
@@ -38,3 +38,12 @@ platform(
         }
         """,
 )
+
+nodejs_binary(
+    name = "tsc_wrapped_with_angular",
```
Comment (devversion) [inline tools/BUILD.bazel]: Unless I miss anything, I think this is no longer used right?

File: packages/compiler-cli/BUILD.bazel [hunk 30]
```diff
@@ -133,6 +133,7 @@ pkg_npm(
     # Do not add more to this list.
     # Dependencies on the full npm_package cause long re-builds.
     visibility = [
+        "//devtools/tools/esbuild:__pkg__",
```
Comment (devversion) [inline packages/compiler-cli/BUILD.bazel]: I see many cases where you change the visibility of the `npm_package` targets. I think this was a leftover from a previous attempt right? If so, can you please try cleaning this up in all of `packages/..`?

**Review by devversion — APPROVED** [review]
LGTM overall. Left a few comments, but overall this looks good to me.

Note that we will need to re-audit how tests are executed, ensuring everything runs as desired, but I don't have sufficient context what tests we want to run as part of the devtools folder. Currently I just see that `devtools/...` runs, but there might be some tests outside Bazel not running I assume.

File: devtools/projects/demo-no-zone/src/BUILD.bazel (hunk 21 above)
Comment (AleksanderBodurri) [reply devtools/projects/demo-no-zone/src/BUILD.bazel:62]: Created a README in the devserver directory to note that it was duped from angular/components

File: .circleci/config.yml (hunk 15 above)
Comment (mgechev) [reply .circleci/config.yml:280]: >I think we need to make a discussion whether we'd want to use Cypress in this repo / have all team members install it, especially since it's just rarely used for the majority of people in the repo, right?

I think we shouldn't have a requirement for all team members to install Cypress unless that's our e2e testing framework by choice. As part of the roadmap, we have migration from Protractor to a different e2e solution. I'd suggest:
1. Not run Cypress in CI for DevTools
2. Merge the PR
3. Migrate to the default e2e testing solution for the repo when we agree on it
4. Enable e2e tests for DevTools in CI

Comment (devversion) [reply .circleci/config.yml:280]: @mgechev that sounds perfect to me 👍

Comment (josephperrott) [reply .circleci/config.yml:280]: I think that makes sense to me as well.  As Paul mentioned, we definitely don't want to pay the cache cost if we don't have to, and I don't want to make every dev install on yarn install it when most wont use it.

File: packages/bazel/src/ng_module/ng_module.bzl [hunk 31]
```diff
@@ -208,6 +208,7 @@ def _ngc_tsconfig(ctx, files, srcs, **kwargs):
     generate_ve_shims = _generate_ve_shims(ctx)
     compilation_mode = _get_ivy_compilation_mode(ctx)
     is_devmode = "devmode_manifest" in kwargs
+
```
Comment (josephperrott) [inline packages/bazel/src/ng_module/ng_module.bzl]: I think we can undo this change.

File: devtools/LICENSE [hunk 32]
```diff
@@ -0,0 +1,21 @@
+The MIT License
```
[the remark this answers is not in the document]
Comment (mgechev) [reply devtools/LICENSE:1]: Sounds good to me.

**Review by josephperrott — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
FYI adding the "cleanup" label since there is a merge conflict with the master branch.

**Comment by AleksanderBodurri** [comment]
> FYI adding the "cleanup" label since there is a merge conflict with the master branch.

Resolved, CI is running now

**Comment by mgechev** [comment]
@AleksanderBodurri resolved the merge conflicts. Just removed the label. Should be ready to go 🚀

**Comment by josephperrott** [comment]
This PR was merged into the repository by commit 8fb3aed.
